conans/test/integration/layout/test_local_commands.py

```diff
@@ -0,0 +1,269 @@
+    conan_file += """
+    def configure(self):
+        self.layout.build.folder = "my_build"
+    def build(self):
+        tools.save("build_file.dll", "bar")
+    """
+    client.save({"conanfile.py": conan_file})
+    client.run("install . -if=common")
+    client.run("build . -if=common -bf=common")
+    dll = os.path.join(client.current_folder, "common", "my_build", "build_file.dll")
+    assert os.path.exists(dll)
+
+
+def test_local_source():
+    """If we configure a source folder in the layout, the downloaded files in a "conan source ."
+    go to the specified folder: "my_source"
+    """
+    # FIXME: The configure is not valid to change the layout, we need the settings and options
+    #        ready
+    client = TestClient()
+    conan_file = str(GenConanfile().with_import("from conans import tools"))
+    conan_file += """
+    def configure(self):
+        self.layout.source.folder = "my_source"
+
+    def source(self):
+        tools.save("downloaded.h", "bar")
+    """
+    client.save({"conanfile.py": conan_file})
+    client.run("install . -if=my_install")
+    # FIXME: This should change to "source ." when "conan source" computes the graph
+    client.run("source . -if=my_install")
+    header = os.path.join(client.current_folder, "my_source", "downloaded.h")
+    assert os.path.exists(header)
+
+
+def test_local_source_change_base():
+    """If we configure a source folder in the layout, the souce files in a "conan source ."
+    go to the specified folder: "my_source under the modified base one "all_source"
+    """
+    # FIXME: The configure is not valid to change the layout, we need the settings and options
+    #        ready
+    client = TestClient()
+    conan_file = str(GenConanfile().with_import("from conans import tools"))
+    conan_file += """
+    def configure(self):
+        self.layout.source.folder = "my_source"
+
+    def source(self):
+        tools.save("downloaded.h", "bar")
+    """
+    client.save({"conanfile.py": conan_file})
+    client.run("install . -if=common")
+    client.run("source . -if=common -sf=common")
+    header = os.path.join(client.current_folder, "common", "my_source", "downloaded.h")
+    assert os.path.exists(header)
+
+
+def test_local_package():
+    """If we configure a package folder in the layout, the packaged files in a "conan package ."
+    go to the specified folder: "my_package" and takes everything from the correct build and
+    source declared folders
+    """
+    # FIXME: The configure is not valid to change the layout, we need the settings and options
+    #        ready
+    client = TestClient()
+    conan_file = str(GenConanfile().with_import("from conans import tools"))
+    conan_file += """
+    no_copy_source = True
+
+    def configure(self):
+        self.layout.source.folder = "my_source"
+        self.layout.build.folder = "my_build"
+        self.layout.package.folder = "my_package"
```

Comment:
This is confusing, or at least unexpected. When I create the package in the local cache, it puts the files under a "my_package" folder in the cache, so the files are not found (the set(CONAN_LIB_DIRS_MYPKG ) is empty in the ``conanbuildinfo.cmake``.

I am not sure how this folder would work in the cache, or what benefit/feature it will implement.

Reply:
I pushed some changes fixing the location of the generated conaninfo and testing it.
About changing the package folder in the layout, the test, by modifying the `cpp_info`, is testing that the package "usable". That part will be more automatic when the info belongs to the package.

The benefit is more clearly seen with the editable package I think. But nevertheless, you could use the `self.in_local_cache` to declare a different layout in the cache. I added a test.

Reply:
Check last commit (https://github.com/conan-io/conan/pull/8246/commits/95ec8bce6cb2da0bdaffb211dd80bb9bafb7b4d5) where the cache, by default, is not following the layout for the package. I'm not convinced either about that. WDYT?